concurrency/pool.go

```diff
@@ -0,0 +1,85 @@
+package concurrency
+
+import (
+	"context"
+	"sync"
+
+	"github.com/grafana/dskit/multierror"
+)
+
+// WorkerPool ensures that for any number of concurrent ForEachNotInFlight calls each with any number of tokens only up to numWorkers f
```

Comment:
I'm not sure I understand this. What is `ForEachNotInFlight` Naively I imagine a worker pool to only have max `n` concurrent executions of `f`. Is this the case here as well?

Reply:
the sentence tries to explain exactly this. Apparently it fails to explain it 😄 I will try to break it into multiple sentences. maybe that makes it easier to understand

Reply:
I tried rewording it in c3ae58ed8f8df0cb9e838b80e2ed8d23e3084013